File: backend/routes/weather.py
from flask import Blueprint, request, jsonify
import requests
import os
from utils import generate_humor
from flask_jwt_extended import create_access_token, jwt_required, get_jwt_identity

weather_bp = Blueprint("weather", __name__)

OPENWEATHER_API_KEY = os.getenv("OPENWEATHER_API_KEY")

@weather_bp.route("/get_weather", methods=["GET"])
@jwt_required()
def get_weather():
    city = request.args.get("city")
    if not city:
        return jsonify({"error": "Please provide a city name"}), 400

    # Call OpenWeather API
    url = f"http://api.openweathermap.org/data/2.5/weather?q={city}&appid={OPENWEATHER_API_KEY}&units=metric"
    response = requests.get(url)
    
    if response.status_code != 200:
        return jsonify({"error": "City not found"}), 404
    
    data = response.json()
    
    weather_info = {
        "city": city,
        "temp": data["main"]["temp"],
        "description": data["weather"][0]["description"],
        "humidity": data["main"]["humidity"],
        "wind_speed": data["wind"]["speed"],
        "feels_like": data["main"]["feels_like"],
        "clouds": data["clouds"]["all"]
    }

    # Generate humor
    funny_comments = generate_humor(weather_info)

    return jsonify({
    "city": city,
    "temperature": weather_info["temp"],
    "description": weather_info["description"],
    "humidity": weather_info["humidity"],
    "wind_speed": weather_info["wind_speed"],
    "feels_like": weather_info["feels_like"],
    "funny_temperature": funny_comments["temperature"],
    "funny_humidity": funny_comments["humidity"],
    "funny_wind": funny_comments["wind"],
    "cloudiness": weather_info["clouds"],
    "funny_clouds": funny_comments["clouds"]
})

File: frontend/js/weather.js
// Extract OAuth token from URL and store it
const urlParams = new URLSearchParams(window.location.search);
const tokenFromUrl = urlParams.get('token');

if (tokenFromUrl) {
    localStorage.setItem('token', tokenFromUrl);
    // Clean up URL (remove token from address bar)
    window.history.replaceState({}, document.title, window.location.pathname);
    console.log('✅ Token saved from OAuth!');
}
function getWeather() {
    const city = document.getElementById("city").value.trim();
    if (!city) return alert("Please enter a city name!");

    const token = localStorage.getItem("token"); // get JWT

    fetch(`http://127.0.0.1:5000/get_weather?city=${city}`, {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            "Authorization": `Bearer ${token}` // send JWT for authentication
        }
    })
        .then(response => {
            if (!response.ok) {
                if (response.status === 401) {
                    alert("Unauthorized! Please login again.");
                    window.location.href = "login.html";
                }
                throw new Error("HTTP error " + response.status);
            }
            return response.json();
        })
        .then(data => {
            const container = document.getElementById("result");
            container.style.display = "block";
            container.innerHTML = `
        <div class="metric"><strong>City:</strong> ${data.city}</div>
        <div class="metric"><strong>Weather:</strong> ${data.description}</div>
        <div class="metric"><strong>Temperature:</strong> ${data.temperature}°C</div>
        <div class="joke">${data.funny_temperature}</div>
        <div class="metric"><strong>Humidity:</strong> ${data.humidity}%</div>
        <div class="joke">${data.funny_humidity}</div>
        <div class="metric"><strong>Wind Speed:</strong> ${data.wind_speed} km/h</div>
        <div class="joke">${data.funny_wind}</div>
        <div class="metric"><strong>Cloudiness:</strong> ${data.cloudiness}%</div>
        <div class="joke">${data.funny_clouds}</div>
      `;
        })
        .catch(error => {
            console.error("Error:", error);
            alert("Failed to fetch weather data. Make sure your backend is running!");
        });
}
